src/corenetwork/lteip/LteIp.cc

```diff
@@ -132,195 +135,226 @@ void LteIp::endService(cPacket *msg)
+        IInterfaceTable *interfaceTable = getModuleFromPar<IInterfaceTable>(
+                par("interfaceTableModule"), this);
         // TODO: how do we find the LTE interface?
-        src = interfaceTable->getInterfaceByName("wlan")->ipv4Data()->getIPAddress();
+        src = interfaceTable->getInterfaceByName("wlan")->getIpv4Address();
         EV << "Local address used: " << src << endl;
+    } else {
+        EV << "Source address in control info from transport layer " << src
+                  << endl;
     }
-    else
-    {
-        EV << "Source address in control info from transport layer " << src << endl;
-    }
-    datagram->setSrcAddress(src);
+    header->setSrcAddress(src);
 
     // set other fields
-    datagram->setTypeOfService(ipControlInfo->getTypeOfService());
-    datagram->setTransportProtocol(ipControlInfo->getProtocol());
+    header->setTypeOfService(ipControlInfo->getTypeOfService());
+    // TODO: header immutable? may be dropped
+    datagram->popAtFront<Ipv4Header>();
+    datagram->insertAtFront(header);
+    //ipControlInfo->setTransportProtocol(ipControlInfo->getProtocol());
 
     // fragmentation fields are correctly filled, but not used
-    datagram->setIdentification(curFragmentId_++);
-    datagram->setMoreFragments(false);
-    datagram->setDontFragment(ipControlInfo->getDontFragment());
-    datagram->setFragmentOffset(0);
+    ipControlInfo->setIdentification(curFragmentId_++);
+    ipControlInfo->setMoreFragments(false);
+    ipControlInfo->setDontFragment(ipControlInfo->getDontFragment());
+    ipControlInfo->setFragmentOffset(0);
 
     // ttl
     short ci_ttl = ipControlInfo->getTimeToLive();
-    datagram->setTimeToLive((ci_ttl > 0) ? ci_ttl : defaultTimeToLive_);
+    ipControlInfo->setTimeToLive((ci_ttl > 0) ? ci_ttl : defaultTimeToLive_);
 
     //** Add control info for stack **
     unsigned short srcPort = 0;
     unsigned short dstPort = 0;
-    int headerSize = IP_HEADER_BYTES;
-
-    switch (ipControlInfo->getProtocol())
-    {
-        case IP_PROT_TCP:
-            inet::tcp::TCPSegment* tcpseg;
-            tcpseg = check_and_cast<inet::tcp::TCPSegment*>(transportPacket);
-            srcPort = tcpseg->getSrcPort();
-            dstPort = tcpseg->getDestPort();
-            headerSize += tcpseg->getHeaderLength();
-            break;
-        case IP_PROT_UDP:
-            inet::UDPPacket* udppacket;
-            udppacket = check_and_cast<inet::UDPPacket*>(transportPacket);
-            srcPort = (unsigned short) udppacket->getSourcePort();
-            dstPort = (unsigned short) udppacket->getDestinationPort();
-            headerSize += UDP_HEADER_BYTES;
-            break;
+    B headerSize = B(IP_HEADER_BYTES);
+
+    auto protocolId = ipControlInfo->getProtocolId();
+    if (protocolId == IP_PROT_TCP) {
+        inet::tcp::TcpHeader* tcpseg;
+        tcpseg = check_and_cast<inet::tcp::TcpHeader*>(transportPacket);
```

Comment:
I think transportPacket cannot be cast to a TcpHeader, because it's a packet now a chunk.

Reply:
Yes, this part was not correctly implemented - is not used in the existing simulation scenarios and therefore this was not discovered. Thanks for spotting the problem! We fixed this one and similar ones in the same file.